### PR TITLE
 Prevent over long nonces in ChaCha20-Poly1305

### LINKED ISSUE #8345: chacha20-poly1305: fails to detect tampering, where IV > 12 and IV <= 16 (CHACHA_CTR_SIZE)
I have been fuzzing OpenSSL's implementation of several AEAD ciphers in OpenSSL 1.1.1a.

For `chacha20-poly1305`, OpenSSL supports an IV, where IV >= 1 byte and IV <= 16 bytes (`CHACHA_CTR_SIZE`), with the default IV being 12 bytes or 96-bits.

However, fuzzing revealed test vectors where 4 bytes of the IV range can be modified by an attacker, without `EVP_CipherFinal_ex()` detecting the corruption.

This means that where authenticated messages use `chacha20-poly1305` as AEAD cipher, and where a 128-bit nonce is included in the authenticated message (instead of the default 96-bit nonce), an attacker can modify 4 bytes of the authenticated message without detection.

Here are test vectors showing IVs for which these collisions are possible:

```
algorithm=chacha20-poly1305
enc=0 (decrypting...)
key=5afd42de4b3da741009ce0f6e8fdf9d3e05142d4dd4998b19c192379523a7c62 (32 bytes)
iv=bab79b546a22cb759a2a7dbcf033 (14 bytes) [corrupt at offset 0, original=b9b79b546a22cb759a2a7dbcf033]
ciphertext=1bc91bf03ef31fa4df81341172b24241697d1313 (20 bytes)
aad=130afce35153a185dd55a5315fa03fdb67ce4a053f58 (22 bytes)
tag=5e202b695689a6d0bbe57aed922bac61 (16 bytes)
TAMPERING NOT DETECTED

algorithm=chacha20-poly1305
enc=0 (decrypting...)
key=c4a70a7cd67c6e0499a323fe85288b946d6d1b5103c6f9b50f5b0dfb92f9edb9 (32 bytes)
iv=765935269f6e019f8ae5aae8353df7 (15 bytes) [corrupt at offset 1, original=765835269f6e019f8ae5aae8353df7]
ciphertext=a59d24 (3 bytes)
aad=a89b005817a4103bc034b004c3afc5c534e85d3650513640df36 (26 bytes)
tag=ced3b91249e3a0ecb3f39b83f6fcd9 (15 bytes)
TAMPERING NOT DETECTED

algorithm=chacha20-poly1305
enc=0 (decrypting...)
key=4e090c4887311123b914fabebd4ed066ce5ee6bc4695d20e34f19b4b99ae1bf7 (32 bytes)
iv=fe97b3cd475439553a585d5982fb0aed (16 bytes) [corrupt at offset 2, original=fe97b2cd475439553a585d5982fb0aed]
ciphertext=95a448b9f3499e1d9f1cee55cede31ecf09d1e07a67b98d9ba92afea1e (29 bytes)
aad=3cdc133a0e16d2df9e0469 (11 bytes)
tag=c2650e0f8db4cd484defa126d108d4 (15 bytes)
TAMPERING NOT DETECTED

algorithm=chacha20-poly1305
enc=0 (decrypting...)
key=e3c30b4556809968baf0277bfdd716a2059cc8d1652310130e49f845318ccf58 (32 bytes)
iv=2309bdf81e51fd18661466a4ccaf29ee (16 bytes) [corrupt at offset 3, original=2309bdf71e51fd18661466a4ccaf29ee]
ciphertext=81bf678661c6d8635aef310f41fed675e27efd54a60cf070710d9f8c43 (29 bytes)
aad=(0 bytes)
tag=52d7f5c6a131b7f5d236805afa8d (14 bytes)
TAMPERING NOT DETECTED
```

I understand that 96-bit nonces are the default, but accepting 128-bits is dangerous if it allows the possibility of tampering with the cipher's context.

### PR DESCRIPTION
**Pull request by mattcaswell**
ChaCha20-Poly1305 is an AEAD cipher, and requires a unique nonce input for
every encryption operation. RFC 7539 specifies that the nonce value (IV)
should be 96 bits (12 bytes). OpenSSL allows a variable nonce length and
front pads the nonce with 0 bytes if it is less than 12 bytes. However it
also incorrectly allows a nonce to be set of up to 16 bytes. In this case
only the last 12 bytes are significant and any additional leading bytes are
ignored.

It is a requirement of using this cipher that nonce values are unique.
Messages encrypted using a reused nonce value are susceptible to serious
confidentiality and integrity attacks. If an application changes the
default nonce length to be longer than 12 bytes and then makes a change to
the leading bytes of the nonce expecting the new value to be a new unique
nonce then such an application could inadvertently encrypt messages with a
reused nonce.

Additionally the ignored bytes in a long nonce are not covered by the
integrity guarantee of this cipher. Any application that relies on the
integrity of these ignored leading bytes of a long nonce may be further
affected.

Any OpenSSL internal use of this cipher, including in SSL/TLS, is safe
because no such use sets such a long nonce value. However user
applications that use this cipher directly and set a non-default nonce
length to be longer than 12 bytes may be vulnerable.

Fixes #8345

### CODE REVIEW

**Comment by mattcaswell** [comment]
Note: CVE number will be added to commit message before commit.

**Comment by jorangreef** [comment]
Should the docs also be updated by this commit?

https://www.openssl.org/docs/man1.1.1/man3/EVP_CipherInit.html

```
ChaCha20-Poly1305
The following ctrls are supported for the ChaCha20-Poly1305 AEAD algorithm.

EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_SET_IVLEN, ivlen, NULL)
Sets the nonce length. This call can only be made before specifying the nonce.
If not called a default nonce length of 12 (i.e. 96 bits) is used.
The maximum nonce length is 16 (CHACHA_CTR_SIZE, i.e. 128-bits).
```

The last line should change?

```
The maximum nonce length is 12 (CHACHA20_POLY1305_MAX_IVLEN, i.e. 96-bits).
```

**Comment by mattcaswell** [comment]
> The last line should change?

Ah! Yes. Thanks - good point. Will fix.

**Comment by mattcaswell** [comment]
New commit added fixing the documentation. That commit will not be back-ported to 1.1.0 since the offending sentence does not appear in the documentation in that version.

File: doc/man3/EVP_EncryptInit.pod [hunk 1]
```diff
@@ -479,7 +479,9 @@ The following I<ctrl>s are supported for the ChaCha20-Poly1305 AEAD algorithm.
 
 Sets the nonce length. This call can only be made before specifying the nonce.
 If not called a default nonce length of 12 (i.e. 96 bits) is used. The maximum
-nonce length is 16 (B<CHACHA_CTR_SIZE>, i.e. 128-bits).
+nonce length is 12 bytes (i.e. 96-bits). If a nonce of less than 12 bytes is set
+then the nonce is automatically front padded with 0-bytes until the nonce is 12
```
Comment (levitte) [inline doc/man3/EVP_EncryptInit.pod]: Will everyone understand "front padded"? I'm wondering if that might need a slight rephrase...

Comment (mattcaswell) [reply doc/man3/EVP_EncryptInit.pod]: Hmm, "front padded" makes sense to me. I could talk about the "leading bytes" being set to 0 if you prefer, but I'm not sure if it's any clearer. Unless you have a suggestion?

Comment (levitte) [reply doc/man3/EVP_EncryptInit.pod]: "then the nonce is automatically padded with leading 0 bytes to make a 12 byte nonce"

Comment (mattcaswell) [reply doc/man3/EVP_EncryptInit.pod]: Fixup commit pushed including that wording.

**Comment by mattcaswell** [comment]
Pushed. Thanks.

**Comment by davidben** [comment]
Variable-length ChaCha20-Poly1305 nonces appear to be a non-standard OpenSSL extension. They aren't covered by the security considerations in RFC 7539. What exactly are this extension's security properties? I don't see this documented anywhere.

CVE-2019-1543 suggests OpenSSL intended to promise that passing two different nonce values, assuming they were accepted, would not break the cipher. What exactly are the conditions? Right now an empty nonce and a nonce of all zeros still collide. Is that an OpenSSL bug, or is this considered misuse of the OpenSSL extension? 

(I would suggest not adding such extensions. If the specification says the AEAD takes 12-byte nonces, only take 12-byte nonces. Were the primitive designed to take variable-length nonces, it would have encoded the length of the nonce somewhere.)

**Comment by mattcaswell** [comment]
The documentation in this PR is supposed to make it clear what happens if an IV with less the 12 bytes is provided, i.e. it is front padded with 0 bytes to make it as 12 byte nonce, e.g. a single byte of value "1" is exactly identical to 11 bytes of 0s followed by "1".

**Comment by davidben** [comment]
Right, that tells you the implementation, but not the caller's obligations. One could have documented 16-byte nonces as valid but truncated, but that wouldn't exactly fit the usual rules for an AEAD (nonces that differ only in the final byte are okay), so there is a clearly an implicit security criteria here.

I would recommend doing both of the following:
* Explicitly document that the caller must not mix multiple nonce lengths for a given key. This non-standard extension is only safe assuming all nonces have the same length.
* Deprecate this non-standard extension and document that callers should only use the standard 12-byte nonce length.

**Pull request closed without merging**